### ISSUE TITLE
[$500] `Waiting for you to add a bank account` appears after adding deposit only bank account in OldDot

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#42105]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.73-0
**Reproducible in staging?:** yes
**Reproducible in production?:** Can't check, requires real credentials
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @mattmoore
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1715610623341999

## Action Performed:

1. Create an expense report as an employee and submit
2. Have it approved on a New Expensify workspace chat and Pay with expensify
3. Go to OldDot and add a deposit only account
4. Return to the expense report on New Expensify
5. The `add bank account` button should't be there

Great comment here outlining what might be wrong [here](https://github.com/Expensify/App/issues/42105#issuecomment-2153687160). 
## Expected Result:
The `add bank account` button shouldn't be there
## Actual Result:
The `add bank account` button still there  and refreshing the page doesn't help

## Workaround:
unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/38435837/cd35ee43-08eb-4c26-bdb6-7e744afdf04f

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01a5f3ad3418d9fb9d</li>
        <li>Upwork Job ID: 1791214044572852224</li>
        <li>Last Price Increase: 2024-06-07</li>
    </ul>
</details>

**Comment by sakluger** [comment]
I will triage this tomorrow.

**Comment by sakluger** [comment]
I triaged and added the External label, but I will also be OOO through May 31. @adelekennedy can you please manage the issue until then? Thanks!

**Comment by adelekennedy** [comment]
waiting for proposals!

**Comment by ikevin127** [comment]
**`Overdue` notes:** Looking for proposals, bumped in #expensify-open-source [Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1716230052317729).

**Comment by adelekennedy** [comment]
Still waiting for proposals here!

**Comment by adelekennedy** [comment]
Still no bites @ikevin127 do you think this is because it's difficult to reproduce?

**Comment by ikevin127** [comment]
@adelekennedy Yes, I think that's part of it. I think another part is that the issue **might** involve both BE and FE changes and because of the current bounty, contributors are not willing to take on fixing it.

I would suggest increasing the bounty and if that doesn't get any bites, I'd make it `Internal`.

**Comment by adelekennedy** [comment]
upping the bounty!

**Comment by ikevin127** [comment]
Bumped on [Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1717186047110379) **#expensify-open-source** as well.

**Comment by b4s36t4** [comment]
Can anyone help me with a video on re-producing this? I tried but wasn't able to get the `Add Bank account` button.

**Comment by ikevin127** [comment]
I tried to reproduce and got confused, wasn't able to as the steps are not that clear and also the video does not show the progression of the steps. Can somebody provide clearer testing steps with video from start to finish including both OD and ND sides ?

@m-natarajan @adelekennedy

**Comment by sakluger** [comment]
Thanks @adelekennedy for watching things while I was out! I'll unassign you now.

**Comment by sakluger** [comment]
@ikevin127 can you clarify which part of the reproduction steps are unclear? Are you unsure about which test accounts take which actions?

**Comment by ikevin127** [comment]
Sure, I think clarifications to the following points would help everybody who's trying to reproduce this issue and work towards finding a solution.

> 1. Create an expense report as an employee and submit

**To clarify:**
1. Create an expense report where ? OD or ND
2. As an employee of what type of workspace ? (Fee, Collect, etc.)
3. Does this workspace have an admin (Owner) and the employee is a **Member** of this workspace or an **Admin** ?
4. Does this workspace have a BA added already ? This would mean that the workspace Owner / Admins are able to issue payments via `Pay with expensify`.

I think there should be more and clearer steps here instead of multiple (assumed) actions crammed-up into one, since anybody should be able to reproduce the issue on staging by simply following the steps, but the current steps assume (a lot of) prior knowledge of the app.

**Note:** I also think this is part of the reason why this issue didn't get any love from the community 💔 

> 2. Have it approved on a New Expensify workspace chat and Pay with expensify

**To clarify:**
1. Have it approved by whom ? The **Owner** of the workspace or any other **Admin** which has the permissions to Approve and `Pay with expensify` ?
2. When whoever approved tries to `Pay with expensify`, does this mean that the workspace needs to have a BA added in order to `Pay with expensify` ? Or the individual user that Approved and wants to `Pay with expensify` needs to have BA added individually ?

> 3. Go to OldDot and add a deposit only account

**To clarify:**
1. This assumes that the account of the workspace employee which is logged-in on OD, did NOT have a BA added prior to this in their **Settings > Account > Payments > Bank account** section ?

**Regarding steps 4 & 5, to clarify:**
1. Assuming that the user was already seeing `Add bank account` on ND after performing step (2.), we're expecting the `Add bank account` button to dissapear in real-time via pusher events ?
2.  Would we be ok with the button going away if the user performs page refresh or re-visits the report ?

If answer is **yes** for question (1.) and **no** for (2.), since this is already not the case (proven by this issue), we might only need BE changes here if the FE already works as expected.

##

Once I have clarifications to these and I am able to reproduce, I will be able to tell whether this is BE only or BE / FE or FE only and potentially will propose to fix this myself if nobody else figures it out in the meantime.

> Can somebody provide clearer testing steps with video from start to finish including both OD and ND sides ?

The reason why I asked for this type of video is because it would clear-up a lot of the questions above, visually, which I think it's much more easier to understand / reproduce.

> [!note]
> I also checked-out the [Slack bug report](https://expensify.slack.com/archives/C049HHMV9SM/p1715610623341999) and I noticed that the steps are a bit more clear but I was still not able to reproduce as there are a lot of assumptions (see clarification questions above) which probably led me the wrong way in reproducing on my side.

cc @sakluger

**Comment by sakluger** [comment]
Thanks, @ikevin127 for clarifying. I'm happy to clarify, although I think that most of the clarifications you ask for unnecessary. Maybe it's because of my familiarity with Expensify in general, but these reproduction steps seem clearer to me than many others I've seen in the past.

**Step 1 (Create an expense report as an employee and submit):**
1. Create an expense report where ?
    - You can generally assume that all actions are taken in NewDot unless stated otherwise.
3. As an employee of what type of workspace?
    - Collect. Free workspaces don't have approval workflows, and we've also just deprecated the free workspace.
4. Does this workspace have an admin (Owner) and the employee is a Member of this workspace or an Admin?
    - Yes - every workspace has an admin, and in order to submit an expense report on a workspace you must be a member of that workspace.
5. Does this workspace have a BA added already? This would mean that the workspace Owner / Admins are able to issue payments via Pay with expensify.
    - Since a later step asks you to add a bank account, we can assume that the submitter does not have a previously added bank account.

**Step 2 (Have it approved on a New Expensify workspace chat and Pay with expensify):**
1. Has it been approved by whom? The workspace owner or any other Admin who has the permissions to Approve and Pay with Expensify?
    - Either is fine. The error is related to what the submitter sees, so as long as it's submitted to an admin who can approve and pay, it should be reproduced.
2. When whoever approved tries to Pay with Expensify, does this mean that the workspace needs to have a BA added in order to Pay with Expensify? Or does the individual user who Approved and wants to Pay with Expensify need to have a BA added individually?
    - The approver/admin would need a verified business bank account added to their workspace

**Step 3 (Go to OldDot and add a deposit only account):**
1. This assumes that the account of the workspace employee, which is logged in on OD, did NOT have a BA added prior to this in their Settings > Account > Payments > Bank account section?
    - Correct, the submitter did not have a deposit bank account added before submitting the report.

**Step 4/5 (Return to the expense report on New Expensify; The add bank account button shouldn't be there):**

1. Assuming that the user was already seeing Add bank account on ND after performing step (2.), we're expecting the Add bank account button to disappear in real-time via pusher events?
    - I think we do want the buttons to go away, even without a refresh. I'll ask in Slack to verify. Either way, the buttons should definitely go away after a refresh.

Regarding the video, it looks like there is a video that starts from step 3. @m-natarajan are you able to redo the reproduction video to also include steps 1 and 2?

**Comment by ikevin127** [comment]
@sakluger Thanks for clearing things up! I managed to reproduce the issue and find the root cause, which is **strictly BE related** in the sense that when deposit-only BA is added by Employee in OD, there's no pusher event sent to ND side to update the Employee's `userWallet` data which would remove the `Add bank account` button.

In order for the button to disappear, I had to go to **ND Settings > Wallet** which fetched the `userWallet` data, only then -> upon returning back to the report, the `Add bank account` button was gone and the OD bank account was registered as added by ND.

Here's the diff in `userWallet` data before / after adding deposit-only BA on OD and navigating to **ND Settings > Wallet** on Employee's side:

```diff
{
    "availableBalance": 0,
    "bankAccountID": 4987807,
    "currentBalance": 0,
    "currentStep": "",
    "hasAcceptedTerms": false,
    "hasFailedOnfido": false,
    "isPendingOnfidoResult": false,
    "remainingWalletLimit": 49999,
    "state": 0,
    "tier": 1,
    "tierName": "SILVER",
    "walletLimit": 50000,
    "walletLimitEnforcementPeriod": 60,
-   "walletLinkedAccountID": 0,
+   "walletLinkedAccountID": 5011847,
-   "walletLinkedAccountType": ""
+   "walletLinkedAccountType": "bankAccount",
+   "hasPaymentMethod": true
}
```

**Note:** This info should help BE engineer fix this issue.

**Comment by sakluger** [comment]
Awesome, thanks for the investigation and explanation, @ikevin127! I'll mark this one as internal.

**Comment by sakluger** [comment]
I asked for a volunteer in #wave-collect in Slack.

**Comment by ikevin127** [comment]
@sakluger Should I unassign myself as C+ reviewer since this is going to be handled internally ?

**Comment by ikevin127** [comment]
`Internal`: looking for a volunteer.

**Comment by sakluger** [comment]
Added to HOT PICKS on the project board.

**Comment by sakluger** [comment]
No bites yet.

**Comment by sakluger** [comment]
No volunteers.

**Comment by ikevin127** [comment]
Still looking for volunteer.

**Comment by sakluger** [comment]
Still looking.

**Comment by ikevin127** [comment]
Still looking for a volunteer!

**Comment by sakluger** [comment]
No updates.

**Comment by ikevin127** [comment]
Looking for a volunteer 🎲

**Comment by ikevin127** [comment]
Looking for a volunteer 🎲

**Comment by ikevin127** [comment]
Still looking for a volunteer 🎲

**Comment by ikevin127** [comment]
Still looking for an internal volunteer 🎲

**Comment by joekaufmanexpensify** [comment]
@garrettmknight was there something you were looking for me to do for this one? Or just wait until it gets picked up internally? Asking because I see it had a BZ assignee/C+, and then @flodnv  [removed them](https://github.com/Expensify/App/issues/42105#issuecomment-2259091802) a few weeks ago. Not sure if that was intentional

**Comment by flodnv** [comment]
Yes it was intentional, because I agree with the question `was there something you were looking for me to do for this one?`

**Comment by joekaufmanexpensify** [comment]
Got it, sounds good! In that case will unassign for now if there is nothing I need to do here. Please let me know if there is something for BZ to do, and happy to help out!

**Comment by garrettmknight** [comment]
> Or just wait until it gets picked up internally? 
I was thinking of this bit. We're pushing to find a volunteer this week, but I was thinking we had a BZ member assigned to all bugs to help out.

**Comment by joekaufmanexpensify** [comment]
Got it. We definitely always do at the outset. But if the issue is just waiting for an internal engineer to pick it up (and there's nothing to do with contributors/C+), and it's being prioritized based on where it falls in the release schedule, it doesn't feel like there is much benefit to having a BZ remain assigned. 

Not super passionate though if there is a definitive best practice here. Just was curious as I'd seen the prior BZ be removed.

**Comment by tgolen** [comment]
OK, I think I found what needs to happen for this to get fixed.

When a bank account is created on NewDot, there are Onyx updates created from web-expensify [here](https://github.com/Expensify/Web-Expensify/blob/1867a513a0d97a76362614448d44e8bb1a3f5f09/lib/BankAccountAPI.php#L917-L927).

When a bank account is created on OldDot, there are no Onyx updates created from web-secure [here](https://github.com/Expensify/Web-Secure/blob/d54848f5cd0edeb88adbf5424cea842ff6a7b535/lib/BankAccountAPI.php#L980).

Here is the real kicker... we have no code in Web-Secure to build or send out Onyx updates 😦 . That will require a lot of code being copied to Web-Secure in order to support this flow.

**Comment by tgolen** [comment]
Probably the quickest way to make this happen would be to have the Web-Secure API make a request over to the Web-Expensify API which will compile the bank account list Onux update and send it out with a pusher event.

@nkuoch @iwiznia I'm curious to hear what you think about this approach.

**Comment by iwiznia** [comment]
No good answers here... on one hand doing what you propose is adding more complexity, on the other we are not going to copy all the code related to onyx updates and pusher to web-s.

There's another option, which would be to move the API method from secure to web, would that be a huge effort or simple? Remember that web-s was a mistake and is not really needed and we will eventually deprecate it, so moving things to web will need to happen sooner or later.

**Comment by tgolen** [comment]
I briefly thought about moving the entire API command to Web-E, and I'm not sure if the frontend of web-secure is setup to call commands on both the web-e or the web-s API endpoint (at first glance, it doesn't look like there is any setup for doing this [ref](https://github.com/Expensify/Web-Secure/blob/d54848f5cd0edeb88adbf5424cea842ff6a7b535/site/lib/api.js#L38)). I also have been discussing in slack and in person if this issue is really worth doing anything about. The workaround for me seems adequate: after adding your bank account on OldDot, go to the wallet page in NewDot and your new bank account will sync.

**Comment by iwiznia** [comment]
The right solution would be to move it, so my vote is we either we do that or do nothing since I agree the bug is quite minor.

**Comment by garrettmknight** [comment]
Also discussing alternatives in this [thread](https://expensify.slack.com/archives/C036QM0SLJK/p1724165193760889)

**Comment by nkuoch** [comment]
I agree that if we do something, we should move everything from secure to web about the vbba flow. But that might not be super fast to do, and maybe not worth spending time on oldDot when newDot is the new thing.

What we could do though is to create a web api sendBankAccountOnyxUpdates and call it from secure.
Should be easy enough? Or we could also do nothing as the bug is minor.

**Comment by tgolen** [comment]
OK, thanks! I was just talking with @garrettmknight and this is where we landed. I'm going to add a simple log to web-secure when someone adds a vbba and they have `tryNewDot.classicRedirect.dismissed === false`. Then, I'll followup after a month to check the logs to see how many people are trying to do this flow. If it's low, we can continue to do nothing. Otherwise, we try to come up with some better alternative.

**Comment by tgolen** [comment]
# Daily Update
- I created [a Web-S PR](https://github.com/Expensify/Web-Secure/pull/2735) for this
- I created [a followup issue](https://github.com/Expensify/Expensify/issues/422532) to check the logs in a month

### Next Steps
- @grgia Review and merge the PR

#### ETA
- Merged today, Aug 22

**Comment by tgolen** [comment]
This was deployed and can be closed.

**Comment by tgolen** [comment]
# Update
- When I checked the logs today, I found that there were 44 instances of the log being triggered in the last two weeks.
- This seems like a significant amount that we should ideally do something with.
- I think the best solution here is to move the command from Web-S to Web-E and then have Web-E create the proper onyx updates to send to the frontend when a bank account is created on OldDot

### Next Steps
- @tgolen reopen this issue and start looking into moving the command

#### ETA
- Friday, Oct. 4

**Comment by tgolen** [comment]
# Daily Update
- I've created a [Web-E PR](https://github.com/Expensify/Web-Expensify/pull/43632) that adds a new API command, copied from Web-S to create the bank account and send out Onyx updates to the user
- I've created a [Web-S PR](https://github.com/Expensify/Web-Secure/pull/2752) which calls the new Web-E API command and removes the old code

### Next Steps
- @MonilBhavsar Review and merge the Web-E PR
- Once the Web-E PR is deployed to production, I will take the Web-S PR off HOLD

#### ETA
- Wednesday, Sep 25

**Comment by tgolen** [comment]
# Daily Update
- The [Web-E PR](https://github.com/Expensify/Web-Expensify/pull/43632) was reviewed
- I've responded to the reviews and pushed up changes to the PR

### Next Steps
- @MonilBhavsar Continue to review and then merge the PR
- Once the Web-E PR is deployed to production, I will take the Web-S PR off HOLD

#### ETA
- Wednesday, Sep 25

**Comment by tgolen** [comment]
# Daily Update
- The [Web-E PR](https://github.com/Expensify/Web-Expensify/pull/43632) was merged and deployed to staging

### Next Steps
- @tgolen Wait for the Web-E PR to be deployed to production, then take the [Web-S PR](https://github.com/Expensify/Web-Secure/pull/2752) off HOLD
- Get the Web-S PR reviewed and merged

#### ETA
- Friday, Sep 27

**Comment by tgolen** [comment]
# Daily Update
- The [Web-E PR](https://github.com/Expensify/Web-Expensify/pull/43632) was deployed to production
- The [Web-S PR](https://github.com/Expensify/Web-Secure/pull/2752) was taken off hold and merged

### Next Steps
- Wait for the Web-S PR to deploy to production and then close this out

#### ETA
- Thursday, Sep 26

**Comment by tgolen** [comment]
The final PR was deployed to production yesterday so I am going to close this out.
